### ISSUE TITLE
For a backend group, if the first backend gives candinates, how to discard candinates from the second backend?

### ISSUE THREAD

**Issue by netjune** [company-mode/company-mode#806]
For example, set company-backends as following:
```
(setq company-backends '((:separate company-lsp company-dabbrev-code)))
```

If company-lsp gives candinates, then discard candinates from company-dabbrev-code.
If company-lsp doesn't give candinates, then use candinates from company-dabbrev-code.

**Comment by dgutov** [comment]
I'd say: don't group them. Use a simple list.

But then if `company-lsp` works and simply doesn't know any completions in the given context, `company-dabbrev-code` won't be used.

**Comment by netjune** [comment]
Yes, that is the reason why I put them in a group. But then, there are many duplicated candinates, one from company-lsp with annotation and one from company-dabbrev-code with no annotation.

**Comment by jonnycyr** [comment]
If you use the keyword :with , for example: `(company-lsp :with company-dabbrev-code)`, and set company-transformers to be company-sort-by-backend-importance, like so: `(setq company-transformers '(company-sort-by-backend-importance))` then company-lsp will always appear before company-dabbrev-code. 

This will give you close to what you're looking for, in the sense that company-lsp candiates will be groups and put before company-dabbrev-code so you basically won't see dabbrev-code. Unfortunately if there are no company-lsp candidates then company-dabbrev-code will not be used. 

I'm currently trying to figure out a solution and asked a similar question here: https://github.com/company-mode/company-mode/issues/818. I basically want company-php to always show before company-dabbrev-code but if company-php has no candidates, then still show company-dabbrev-code.

**Comment by dgutov** [comment]
Try `(company-lsp company-dabbrev-code :separate)`. You won't need the transformer in this case.

**Comment by dgutov** [comment]
Hmm, sounds like a valid complaint.

**Comment by dgutov** [comment]
But it's not easy to fix that without changing the behavior of `company-complete` (where we do offer completions independent of `company-minimum-prefix-length`).

**Comment by netjune** [comment]
Yes.
It seems that code completion becomes more complicated since language server was used.

**Comment by dgutov** [comment]
Normally, the language server should also provide the completions you expect from `company-dabbrev-code`. So you wouldn't need to group it.

**Comment by netjune** [comment]
I have only used language server for python(pyls) for one month. It didn't work very well. So I had to group them together. I have not used other language server for too long.

Others' use case:
https://emacs-china.org/t/company-lsp-company-dabbrev-code/8362

**Comment by dgutov** [comment]
:man_shrugging:

I haven't tried LSP yet (I hear the Ruby impls are still very alpha), and I don't write Python.
However, LSP completion should include local variables etc. Maybe with some later release of whatever server you're using.

**Comment by Yevgnen** [comment]
Is it make sense to call other backends when `prefix` is not `nil` and `company-candidates` is `nil` automatically when not using group backends? I want to try but I don't know where to put the code... It seems that `company-completion-*-hook`s are not the right places since they are called either 'started' or 'finished/canceled', but not 'no candidates'. 

In the `:async` backend situation, it seems it should be called in some callback function. 

I really like the idea of [`elpy-company-backend`](https://github.com/jorgenschaefer/elpy/blob/9913456bbc386df10c7d3cf48a07860b008e7941/elpy.el#L3640), when the major backend has no completion with a non-`nil` prefix, it will get some candidates from other backends. 

I've tried to modify the callback part of the source code of `company-lsp`, and add `company-dabbrev-code` candidates, it worked as expected.

But I wonder if there's general solution for (sync/async) backends without manual trigger keys.

**Comment by dgutov** [comment]
The idea behind a dispatching backend is indeed attractive because it's supposed to take care of subtle things like said backends returning different prefix'es. Or switching back to an earlier backend when it starts returning candidates again.

> general solution for (sync/async) backends without manual trigger keys

Meaning, `M-x company-other-backend` is not what you want to use?

**Comment by Yevgnen** [comment]
> Meaning, M-x company-other-backend is not what you want to use?

Yes, if auto complete of other backend can start automatically, that would be perfect, just like most editors and 'IDE's. I've tried hard to hook `company-other-backend` into `company-cancel` naively, but there were many issues to solve.

Whether should manually trigger is a personal taste. But many lang-related packages tried to solve this by adding candidates from other backends like `elpy` mentioned above or by `(add-to-list 'company-backends '(company-meghanada :separate company-dabbrev-code))` (from `meghanada`), because returns nothing is at least not a quite good user experience (for who expects **auto** complete should always happen automatically). For me, keep typing and pressing some key like `TAB` is a bit annoy for me. Thus if the solution exists, I think it will also reduce the complexity for package author or user to design `major-mode` specific `company-backends`.

**Comment by dgutov** [comment]
> that would be perfect, just like most editors and 'IDE's

Could you clarify? What other editors/IDEs have multiple backends with automatic failover in the case of no matches?

> because returns nothing is at least not a quite good user experience

"No matches" is a meaningful result as well. You can take it to mean there's no method named like the one you're typing. Automatic failover to "all words in all buffer" removes that understanding.

> I think it will also reduce the complexity for package author

Having done that myself recently to a Ruby-related package, I wouldn't say that implementing local variables support in a backend is all that difficult. And one you're doing it for a particular language, you can return more helpful results than `company-dabbrev-code`.

**Comment by Yevgnen** [comment]
@dgutov  Hi, my English is poor, but I hope I can explain it well.

I don't know whether these editors/IDEs have multiple backends. What I want to say is they always 'try to give candidates if possible automatically'. For example, vscode, it may first try some lsp backend, then try to get candidates like `dabbrev code` and then try to get something from comment words. 

![vs](https://user-images.githubusercontent.com/16749790/53705275-2cbde080-3e5e-11e9-98be-52ab8582326e.gif)

It may use multiple or one backend. But it never require me to press a `TAB` or something else. What I need to do is just keep typing and reminding that 'it should return me something, and if it does I'm happy to choose one of that to reduce my typing.'

> Having done that myself recently for a Ruby-related package, I wouldn't say that implementing local variables support in a backend is all that difficult. And one you're doing it for a particular language, you can return more helpful results than company-dabbrev-code.

I personally think, from the user side, one should only care what the `company-backends` has and then he could get what he expects. 

For example, when `company-backends` is `(elpy-company-backend company-dict)`, he has already benefits from `company-dabbrev-code` since `elpy` implements something special for he. But he typing a `longlongword` may wonder why there's no candidate because he remember that it should in his personal dictionary and `company-dict` follows `elpy-company-backend`. Then he may try `TAB` some times and he needs to tweak `company-backends` with `:separate` keyword and group these backends to have completion automatically. 

In some other modes, the major backend may not implement the `company-dabbrev-code` trick, the user will get no candidates because the major backend returns a non-nil prefix. Then the user may tweak `company-backends` to deal with `company-dabbrev-code` firstly, and repeat 'no dictionary item' problem above. 

It may not that difficult for experienced authors to implement a local variables support in a backend. The implement of different major backend makes completion experience different in different major modes. Some packages may happy to add specific extra `dabbrev code` candidates for the major backend, and some may add `keyword` candidates or maybe some may add `company-semantic` candidates. But you may notice that people keep asking how completion mechanism is work, why there's no candidate, especially they think there should be. Users may need to tweak every local `company-backends` in different major modes to have the same completion experience('try the major first please, no? try some buffer symbols please, then try some English words...').

**Comment by dgutov** [comment]
> What I need to do is just keep typing and reminding that 'it should return me something, and if it does I'm happy to choose one of that to reduce my typing.'

I'm not a fan of this approach personally, and making it work would require a few architectural changes. But thank you for the example and the animation. To be clear, if you backspace over the last several characters back to `long`, does it use the previous backend for completion? One without the "longlong" words.

**Comment by Yevgnen** [comment]
@dgutov It seems yes (Changed back to the non 'dabbrev' one). The candidates from backspacing are 'almost' the same as the ones when inserting new chars.

![i](https://user-images.githubusercontent.com/16749790/54102347-002c3a80-4403-11e9-85ab-1c1e0c7690f9.gif)

**Comment by dgutov** [comment]
Interestingly, `completion-at-point-functions` (and, through it, the `company-capf` backend) already provide such a feature via the `:exclusive key (with value `no`).

It's implemented, however, by calling the respective backend right away to check whether it provides any completions. Which, at the very least, means that the backend must implement its own caching, or else it will result in double the processing, or (if the caching is done inside the external process), double the network calls.

**Comment by isolet** [comment]
Even simple backends like `company-files` can not play well with other backends in all modes. I'd like to *auto* complete file paths whenever wherever writing path like strings (e.g. ~/ or /) in all modes but usually it can not get control and I have to `M-x company-files` .

**Comment by dgutov** [comment]
Some of the problems mentioned here are also described in other open issues, and some have been fixed (e.g. `company-files` is much better now).

So I'm going to close this, but feel free to post if you (someone) disagree with the assessment.
